File: app/src/main/java/com/shengshijie/serialporttest/SerialPortActivity.java
package com.shengshijie.serialporttest;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import com.shengshijie.serialport.SerialPort;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

public class SerialPortActivity extends Activity {

    private SerialPort mSerialPort = new SerialPort();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.console);
        mSerialPort.init(new File("dev/ttyS4"), 115200, 0);
    }

    private byte[] merge(byte[] hand, byte[] tail) {
        if (hand == null) {
            return tail;
        }
        byte[] data3 = new byte[hand.length + tail.length];
        System.arraycopy(hand, 0, data3, 0, hand.length);
        System.arraycopy(tail, 0, data3, hand.length, tail.length);
        return data3;
    }

    private byte[] receiveBytes(InputStream in, int maxLength) throws IOException {
        int total = 0;
        byte[] buffer = new byte[maxLength];
        int current;
        do {
            current = in.read(buffer, total, maxLength - total);
            if (current > 0) {
                total += current;
            }
        } while (total < maxLength && current != -1);
        return buffer;
    }

    @Override
    protected void onDestroy() {
        mSerialPort.destroy();
        super.onDestroy();
    }

    public void start(View view) {
        mSerialPort.receive(new Function1<FileInputStream, Unit>() {
            @Override
            public Unit invoke(FileInputStream fileInputStream) {
                try {
                    if (fileInputStream.available() <= 0) {
                        return null;
                    }
                    byte[] buffer = new byte[1024];
                    buffer = Arrays.copyOfRange(buffer, 0, fileInputStream.read(buffer));
                    if (buffer[0] == 85 && buffer[1] == -86 && buffer[3] == 0x00) {
                        int len = (buffer[4] & 0xff) + ((buffer[5] << 8) & 0xff);
                        if (len > 0) {
                            byte[] result = new byte[len];
                            System.arraycopy(merge(buffer, receiveBytes(fileInputStream, len)), 6, result, 0, len);
                            Log.e("CCC", new String(result));
                            mSerialPort.stopReceive();
                        }
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    return null;
                }
                return null;
            }
        });
    }

    public void stop(View view) {

    }

}

File: app/src/main/java/com/shengshijie/serialporttest/SerialPortActivity.kt
package com.shengshijie.serialporttest

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.view.View
import com.shengshijie.serialport.SerialPort
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.util.*
import kotlin.experimental.and

class SerialPortActivity : Activity() {
    private val mSerialPort = SerialPort()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.console)
        mSerialPort.init(
            File("dev/ttyS4"), 115200, 0
        ) { isSuccess, errorMsg ->
            if (isSuccess) {
                Log.e("CCC", "isSuccess")
            } else {
                Log.e("CCC", errorMsg)
            }
        }
    }

    private fun merge(hand: ByteArray?, tail: ByteArray): ByteArray {
        if (hand == null) {
            return tail
        }
        val data3 = ByteArray(hand.size + tail.size)
        System.arraycopy(hand, 0, data3, 0, hand.size)
        System.arraycopy(tail, 0, data3, hand.size, tail.size)
        return data3
    }

    @Throws(IOException::class)
    private fun receiveBytes(`in`: InputStream?, maxLength: Int): ByteArray {
        var total = 0
        val buffer = ByteArray(maxLength)
        var current: Int
        do {
            current = `in`!!.read(buffer, total, maxLength - total)
            if (current > 0) {
                total += current
            }
        } while (total < maxLength && current != -1)
        return buffer
    }

    override fun onDestroy() {
        mSerialPort.destroy()
        super.onDestroy()
    }

    fun start(view: View?) {
        mSerialPort.receive { fileInputStream: FileInputStream? ->
            try {
                if (fileInputStream!!.available() <= 0) {
                    return@receive
                }
                var buffer = ByteArray(1024)
                buffer = Arrays.copyOfRange(buffer, 0, fileInputStream.read(buffer))
                if (buffer[0] == (85).toByte() && buffer[1] == (-86).toByte() && buffer[3] == (0x00).toByte()) {
                    val len: Int =
                        (buffer[4] and (0xff).toByte()) + (buffer[5].toInt() shl 8 and 0xff)
                    if (len > 0) {
                        val result = ByteArray(len)
                        System.arraycopy(merge(buffer, receiveBytes(fileInputStream, len)), 6, result, 0, len)
                        Log.e("CCC", String(result))
                        mSerialPort.stopReceive()
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                return@receive
            }
        }
    }

    fun stop(view: View?) {}
}